### PR TITLE
feat(plugin): support using custom project & data catalog (major)

### PR DESCRIPTION
**Pull request by airslice**
## Overview

This PR adds support for customize project & data catalog.

## What i have done

- Sidebar config: add config fields for customize backend project & data catalog.
- Data catalog (PC modal & device popup)
  - Show data catalog of plateau & custom independently.
  - Remember filter / added dataset / explanded folders etc. for each data catalog.
  - Hide publish button for plateau dataset if it's a custom project.
  - Add switch tab UI for data catalog on device.
- Sidebar
  - Get project data from both plateau and custom, combine them into one data.
  - Get template data from both plateau and custom, combine them into one data.
  - Disable edit mode for plateau dataset on dataset card if it's a custom project.
  - Disable edit mode for template item of plateau if it's a custom project.
  - Modify hooks of project / template / dataset to fetch / update / etc. to target backend.
  - Share with target backend.  
 
- Basically i kept most of the actions as it was, just modify payload if possible.

- Front end will try to prevent custom project editor update data of plateau by some UI changes, but the most powerful limit will come from the access token.

## Test

Test project [here](https://test.reearth.dev/edit/01h1jgk248aby5r4rxkxm7g0jg).

### CODE REVIEW

File: plugin/web/extensions/sidebar/popups/mobileDropdown/index.tsx [hunk 1]
```diff
@@ -67,9 +81,17 @@ const MobileDropdown: React.FC = () => {
...
+    postMsg({ action: "saveSearchTerm", payload: { searchTerm: value, dataSource: "plateau" } });
   }, []);
 
+  const handleCustomSearch = useCallback(
+    ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
+      setCustomSearchTerm(value);
+      postMsg({ action: "saveSearchTerm", payload: { searchTerm: value, dataSource: "custom" } });
+    },
+    [],
+  );
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/popups/mobileDropdown/index.tsx:93]: Maybe not needed (just pass datasource in to callback)

File: plugin/web/extensions/sidebar/popups/mobileDropdown/Catalog/index.tsx [hunk 2]
```diff
@@ -40,59 +42,132 @@ type Props = {
...
         setCatalog(res);
       });
     }
-  }, [backendURL, catalogProjectName, catalogURL]);
+    const customCatalogBaseUrl = customCatalogURL || customBackendURL;
+    if (customCatalogBaseUrl) {
+      getDataCatalog(customCatalogBaseUrl, customCatalogProjectName).then(res => {
+        setCustomCatalog(res);
+      });
+    }
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/popups/mobileDropdown/Catalog/index.tsx:137]: This should be in a separate useEffect just so we aren't accidentally recalling both when not needed

File: plugin/web/extensions/sidebar/modals/datacatalog/components/index.tsx [hunk 3]
```diff
@@ -54,13 +73,33 @@ const DataCatalog: React.FC = () => {
           expandedFolders={expandedFolders}
           searchTerm={searchTerm}
           filter={filter}
+          dataSource={"plateau"}
+          editable={!isCustomProject}
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/modals/datacatalog/components/index.tsx:77]: Doesn't this make it so plateau is not editable if there is a city (custom) project? I don't think those things are connected.
If the user has inputted all fields for plateau (including a token) this should be editable. Doesn't matter if custom dataset info is also added or not.

File: plugin/web/extensions/sidebar/core/components/mobile/hooks/projectHooks.ts [hunk 4]
```diff
@@ -283,9 +289,16 @@ export default ({
...
-    if (!backendURL || !backendProjectName || fetchedSharedProject.current) return;
+    if (
+      (!isCustomProject && (!plateauBackendURL || !plateuProjectName)) ||
+      (isCustomProject && (!customBackendURL || !customBackendProjectName)) ||
+      fetchedSharedProject.current
+    )
+      return;
     if (projectID) {
       (async () => {
+        const backendURL = isCustomProject ? customBackendURL : plateauBackendURL;
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/core/components/mobile/hooks/projectHooks.ts:300]: I don't think we need isCustomProject. Just prioritize the customBackendURL (`customBackendURL ?? plateauBackendURL`)

File: plugin/web/extensions/sidebar/core/components/mobile/hooks/index.ts [hunk 5]
```diff
@@ -18,20 +18,41 @@ export default () => {
...
+          resData = plateauResData.map(t => ({ ...t, dataSource: "plateau" }));
+        }
+      }
+      if (isCustomProject) {
+        const res = await fetch(
+          `${customBackendURL}/sidebar/${customBackendProjectName}/templates`,
+        );
+        if (res.status === 200) {
+          const customResData = (await res.json()) as Template[];
+          resData = resData.concat(customResData.map(t => ({ ...t, dataSource: "custom" })));
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/core/components/mobile/hooks/index.ts:53]: Can you move adding dataSource in to the api code? If you do it here, then we are processing the datacatalog in multiple places, which will double the processing time)

File: plugin/web/extensions/sidebar/core/components/desktop/hooks/index.ts [hunk 6]
```diff
@@ -230,17 +271,20 @@ export default () => {
...
     templates: fieldTemplates,
     currentPage,
     loading,
     buildingSearch,
-    searchTerm,
+    isCustomProject,
+    customReearthURL,
+    customBackendURL,
+    customBackendProjectName,
+    //
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/core/components/desktop/hooks/index.ts:285]: delete please

File: plugin/web/extensions/sidebar/core/components/content/common/DatasetCard/index.tsx [hunk 7]
```diff
@@ -384,9 +391,11 @@ const DatasetCard: React.FC<Props> = ({
                         onClick={handleTabChange}>
                         公開
                       </Tab>
-                      <Tab id="edit" selected={currentTab === "edit"} onClick={handleTabChange}>
-                        設定
-                      </Tab>
+                      {editable && (
```
Comment (KaWaite) [inline plugin/web/extensions/sidebar/core/components/content/common/DatasetCard/index.tsx:394]: Can you move this to the conditional rendering that includes the 公開 tab? (so `{inEditor && expanded && editable && (<TabWrapper> ....`

Comment (airslice) [reply plugin/web/extensions/sidebar/core/components/content/common/DatasetCard/index.tsx]: Updated.

File: plugin/web/extensions/sidebar/core/components/desktop/hooks/index.ts (hunk 6 above)
Comment (airslice) [reply plugin/web/extensions/sidebar/core/components/desktop/hooks/index.ts]: Updated.

File: plugin/web/extensions/sidebar/popups/mobileDropdown/Catalog/index.tsx (hunk 2 above)
Comment (airslice) [reply plugin/web/extensions/sidebar/popups/mobileDropdown/Catalog/index.tsx:140]: Updated.

File: plugin/web/extensions/sidebar/modals/datacatalog/api/api.ts [hunk 8]
```diff
@@ -64,6 +64,8 @@ type RawRawDataCatalogItem = {
   code: number;
 };
 
+export type DataSource = "plateau" | "custom";
```
[the remark this answers is not in the document]
Comment (airslice) [reply plugin/web/extensions/sidebar/modals/datacatalog/api/api.ts:67]: Updated.

File: plugin/web/extensions/sidebar/popups/mobileDropdown/index.tsx (hunk 1 above)
Comment (airslice) [reply plugin/web/extensions/sidebar/popups/mobileDropdown/index.tsx]: Updated base on talk from slack. Now using single searchTerm and it will be cleared when switch tabs.
